pkg/controller/bootstrap/bootstrapsigner.go

```
@@ -0,0 +1,262 @@
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/metrics"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const (
+	configMapClusterInfo = "cluster-info"
+	kubeConfigKey        = "kubeconfig"
+	signaturePrefix      = "jws-kubeconfig-"
+)
+
+// BootstrapSignerOptions contains options for the BootstrapSigner
+type BootstrapSignerOptions struct {
+
+	// ConfigMapNamespace is the namespace of the ConfigMap
+	ConfigMapNamespace string
+
+	// ConfigMapName is the name fo the ConfigMap
+	ConfigMapName string
+
+	// ConfigMapResynce is the time.Duration at which to fully re-list configmaps.
+	// If zero, re-list will be delayed as long as possible
+	ConfigMapResync time.Duration
+
+	// SecretResync is the time.Duration at which to fully re-list secrets.
+	// If zero, re-list will be delayed as long as possible
+	SecretResync time.Duration
+}
+
+// DefaultBootstrapSignerOptions returns a set of default options for creating a
+// BootstrapSigner
+func DefaultBootstrapSignerOptions() BootstrapSignerOptions {
+	return BootstrapSignerOptions{
+		ConfigMapNamespace: api.NamespacePublic,
+		ConfigMapName:      configMapClusterInfo,
+	}
+}
+
+// BootstrapSigner is a controller that signs a ConfigMap with a set of tokens.
+type BootstrapSigner struct {
+	client       clientset.Interface
+	configMapKey string
+
+	configMaps cache.Store
+	secrets    cache.Store
+
+	// Since we join two objects, we'll watch both of them with controllers.
+	configMapsController *cache.Controller
+	secretsController    *cache.Controller
+}
+
+// NewBootstrapSigner returns a new *BootstrapSigner.
+func NewBootstrapSigner(cl clientset.Interface, options BootstrapSignerOptions) *BootstrapSigner {
+	e := &BootstrapSigner{
+		client:       cl,
+		configMapKey: options.ConfigMapNamespace + "/" + options.ConfigMapName,
+	}
+	if cl != nil && cl.Core().RESTClient().GetRateLimiter() != nil {
```

What happens if `cl` is indeed `nil`? Maybe we should fail fast?